markdown/developers/modules.md

```diff
@@ -762,6 +762,104 @@ To see some more advanced examples of these keys in use see:
 - [Set ext.prefix based on task inputs](https://github.com/nf-core/rnaseq/blob/e049f51f0214b2aef7624b9dd496a404a7c34d14/conf/modules.config#L297)
 - [Set ext.args based on both parameters and task inputs](https://github.com/nf-core/rnaseq/blob/e049f51f0214b2aef7624b9dd496a404a7c34d14/conf/modules.config#L377-L381)
 
+## Guidance on module inputs
+
+There are various ways information that a given software needs can be entered into a module to make the module work.
+We aim to make nf-core modules as flexible as possible, to ensure pipelines are not constrained by 'particularities' of a given module. In otherways, we want to make modules work for pipelines, not pipelines to work for modules.
+Here we will provide some general guidance on how and when to use the various ways of passing information into a module.
+
+The five ways of inserting information into a module are:
+
+- `input:`: a distinct input channel defined by in the module itself
+- `val(meta)`: a common entry in nf-core module tuple-based input channels that is a list of information associated with a given file
+- `ext.args`: defined in `modules.conf` configuration files that is used to insert (normally non-mandatory) custom flags/options to the tools used in the module
+- `ext.prefix`: defined in `modules.conf` configuration files that is used for defining file names (typically based on a `meta.id` element)
+- `ext.<custom>`: defined in `modules.conf, custom 'variables' that can be passed into the module, defined by the writer of the module
+
+In general the priority order is:
+
+- All files should be inserted into a module via `input:` channels
+- All non-file inputs: `args` > `meta` > `ext.custom`
+  - Where custom pipeline-specific `meta` keys can be inserted into an `args`, rather than module-hardcoded `meta` keys
+
+## ext.args
+
+This is the main way to pass non-file inputs (flags, options, strings, etc.) to a tool within a module. Every nf-core module must include an `$args` variable within the command of the tool itself.
+
+Pipeline developers can insert any information stored in their `meta` maps to this variable via a `modules.config` (see below). Users can also add their own additional personal parameters or information to this via a custom `modules.conf` file (at their own risk).
+
+This can be done via closures, and be dynamically constructued using conditions within lists.
+
+```nextflow
+ext.args = {
+  [
+    "-a" ,
+    "-b ${meta.id}",
+    run_mode ? "-c" : ""
+    ].join(' ') }
+```
+
+It is important to note that it is problematic to have something in ` config` that keeps a _module_ working, as configs can be overwritten. Therefore if the _module_ (not the tool!) requires a particular parameter to execute such information can be passed via an `input:` channel (see below). However, a pipeline can make sure that the _tool_ keeps working (i.e., non-file arguments) via their own `modules.conf`.
```

Comment:
> However, a pipeline can make sure that the _tool_ keeps working (i.e., non-file arguments) via their own `modules.conf`. 

This is not true. A custom config completely overrides `ext.args` and if an option is supplied, the onus ends up on the user to make sure the tool keeps functioning.

Reply:
Ah true!